### DISCUSSION TITLE
Fipy for 2D swe

### DISCUSSION THREAD

**Discussion by MehrdadKianiOsh** [usnistgov/fipy#861] [Q&A]
Hello All,

Thanks to all developers for the solver. I am trying to implement the shallow water equations in Fipy and not sure at all that I wrote the equations in the right way. I appreciate if you could help me in writing the equations:

![image_2022-04-28_17-40-34](https://user-images.githubusercontent.com/100420769/167289946-bf95a4a6-ab59-47ca-ae48-fe62c96ec8e9.png)

and my implementation are (knowing that \tau_x=friction*u^2/8 and \tau_y=friction*v^2/2, y_b is the bed topography and all the three variables are defiend as CellVaiable):

![image](https://user-images.githubusercontent.com/100420769/167290051-5ba4eb87-4485-4e7a-8d79-25a457179548.png)

Best,
Mehrdad

**Comment by wd15** [comment]
Hi @MehrdadKianiOsh: there are plenty of issues that I can see as the equations have been implemented above, but let's just focus on the equation for the height for now. To make the implementation more succinct, you will need to defined a vector velocity, `velocity`, and variables for each component of velocity, `xVelocity` and `yVelocity`. This is simply a quirk of using FiPy. The way you've written it above will likely work, but it makes it difficult to read the equations. Using this method, the variable `velocity` will need to be explicitly updated from `xVelocity` and `yVelocity` before you solve the height equation. The velocities are defined with

```
xVelocity = CellVariable(mesh=mesh, value=0.0, hasOld=True)
yVelocity = CellVariable(mesh=mesh, value=0.0, hasOld=True)
velocity = CellVariable(mesh=mesh, rank=1, value=0.0)
```

The `xVelocity` and `yVelocity` variables are updated by solving the velocity equation. `velocity` is upated from `xVelocity` and `yVelocity` using

```
velocity[0] = xVelocity
velocity[1] = yVelocity
```

whenever required. Using this method, the height equation can be written as

```
heightEq = TransientTerm(var=height) + ConvectionTerm(coeff=velocity, var=height) == 0
```

This change can also be applied to the equations for each velocity component.

Another glaring issue is that values shouldn't be outside of the `DiffusionTerm` expression. For example, the diffusion term expression for the x component of velocity should be written,

```
DiffusionTerm(coeff=2 * viscosity * height, var=xVelocity)
```

We can work on it some more once you have these two things cleaned up.

Here are some links that might help you further (fipy + shallow water equations).

 - [Previous question on FiPy mailing list](https://fipy.nist.narkive.com/EFQX6XGS/2d-shallow-water-model-with)
 - [@guyer notebook on shallow water equations with FiPy](https://github.com/guyer/haefner/blob/master/killworth.ipynb)
 - [Using CLAWPACK instead, better for capturing shocks if that's your goal](https://scicomp.stackexchange.com/questions/41060/shock-capturing-methods-for-shallow-water-equations?rq=1)
 - [FiPy 2D Navier Stokes implementation: problem with derivatives in one direction](https://stackoverflow.com/questions/66243908/fipy-2d-navier-stokes-implementation-problem-with-derivatives-in-one-direction)
